### DISCUSSION TITLE
How do I upload images to the DB from Go?

### DISCUSSION THREAD

**Discussion by bogdano** [pocketbase/pocketbase#4851] [Q&A] [answered]
I am playing around with Go and Pocketbase for the first time, and want to build my entire small project using just Go for the backend. The JS/Dart SDKs are well documented, but I couldn't find documentation on how to do this in Go. I have a function which gets National Parks data from the NPS API, and I want to get the image of a national park from the image URL in their API JSON, convert it to .webp, and then save it in my collection. I've created an image .webp field called "images" (only 1 image until I figure this out), but nothing seems to get saved to the images field (but wepb images do get saved to my project root directory):

```				imageURL := park.Images[0].URL
				resp, err := http.Get(imageURL)
				if err != nil {
					log.Fatalf("Error fetching image: %v", err)
				}
				defer resp.Body.Close()
				imageBytes, err := io.ReadAll(resp.Body)
				if err != nil {
					log.Fatalf("Error reading image bytes: %v", err)
				}
				jpg, err := jpeg.Decode(bytes.NewReader(imageBytes))
				if err != nil {
					log.Fatalf("Error decoding image: %v", err)
				}
				// get filename from ImageURL
				filename := imageURL[strings.LastIndex(imageURL, "/")+1:] + ".webp"
				log.Printf("Filename: %s", filename)
				webpImage, err := os.Create(filename)
				if err != nil {
					log.Fatalf("Error creating webp image: %v", err)
				}
				defer webpImage.Close()
				options, err := encoder.NewLossyEncoderOptions(encoder.PresetDefault, 75)
				if err != nil {
					log.Fatalf("Error creating lossy encoder options: %v", err)
				}
				if err := webp.Encode(webpImage, jpg, options); err != nil {
					log.Fatalf("Error encoding webp image: %v", err)
				}
				record.Set("images", webpImage)
				if err := app.Dao().SaveRecord(record); err != nil {
					return c.String(http.StatusInternalServerError, err.Error())
				}
```

I would really appreciate somebody's input, thank you!

Also, Pocketbase seems like a fantastic project!

edit: I am using https://github.com/kolesa-team/go-webp to encode to .webp

**Comment by ganigeorgiev** [comment, marked as answer]
This will be simplified in the near future as part of the [ongoing refactoring](https://github.com/pocketbase/pocketbase/discussions/4355) but for now the easiest way to upload a file programmatically is to use the [`RecordUpsert` form](https://pocketbase.io/docs/go-records/#create-new-record-with-data-validations):


```go
form := forms.NewRecordUpsert(app, record)

// load regular data
// (or form.LoadRequest(r, ""))
form.LoadData(map[string]any{
    "title": "Lorem ipsum",
    "active": true,
    "someOtherField": 123,
})

// manually upload file(s)
f0, _ := filesystem.NewFileFromBytes(yourBytesData, "image.png")
f1, _ := filesystem.NewFileFromPath("/path/to/file1")
f2, _ := filesystem.NewFileFromUrl(context.TODO(), "https://example.com/image.png") // you can replace the context with one that has timeout
// ... for all possible factories see https://pkg.go.dev/github.com/pocketbase/pocketbase/tools/filesystem#File
form.AddFiles("images", f0, f1, f2)

// validate and save
form.Submit()
```

Also note that webp thumbs are currently not supported - https://github.com/pocketbase/pocketbase/issues/825.

**Reply by bogdano** [reply]
...I can't believe how quickly you responded to me.

Thank you so much! I am implementing this right now

**Reply by bogdano** [reply]
@ganigeorgiev I'm sorry to bump this after closing the question, but is there a way to leverage the CreateThumbs function to resize images BEFORE saving them to the collection? I ask because the NPS API has no consistency or web-readiness when it comes to their images-- some are kb in size, some are ~13mb in size-- and I would like to compress them before saving.

**Reply by ganigeorgiev** [reply]
No, we currently don't expose such functionality and you'll have to create your own helper (note that it may differ depending on what format you are targeting and the type of compression you want).

PocketBase thumbs are generated lazily on first access and they require the original to be present.

**Reply by bogdano** [reply]
Again, I have to say that I've never had this experience on a large OSS project where a maintainer responds so quickly, succinctly, and helpfully. I figured I would share my simple helper function which I've just tested, and it works.

Checking https://github.com/pocketbase/pocketbase/blob/v0.20.2/tools/filesystem/file.go , and using https://github.com/nfnt/resize , this simple helper function works with JPEGs (which is fine for my usecase right now):

```
func downloadAndResizeImage(url string, maxWidth uint) ([]byte, error) {
	// Get the image from the URL
	response, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()

	// Decode the downloaded image
	img, _, err := image.Decode(response.Body)
	if err != nil {
		return nil, err
	}

	// Resize the image if it's wider than the maxWidth
	if img.Bounds().Dx() > int(maxWidth) {
		img = resize.Resize(maxWidth, 0, img, resize.Lanczos3)
	}

	// Encode the image back to a byte slice as JPEG
	buf := new(bytes.Buffer)
	err = jpeg.Encode(buf, img, nil)
	if err != nil {
		return nil, err
	}
	log.Printf("Image resized: %v", url)
	return buf.Bytes(), nil
}
```
The output of this function goes into the `NewFileFromBytes(b []byte, name string)` Pocketbase function.

Thank you again, @ganigeorgiev, you legend.
